### DISCUSSION TITLE
rp2040: MemoryError: memory allocation failed, allocating 2072 bytes

### DISCUSSION THREAD

**Discussion by Flipje1955** [micropython/micropython#12564] [RP2040 / Pico] [answered]
Direct question: is there a practical limit as to the size of a MP program ? Or that question phrased differently: does MP need a minimum of  storage to be able to analyse / run a program and what is that minimum ?

Background: when creating a by now sizable program of some 2000 lines (incl comments), I get an error message "MemoryError: memory allocation failed, allocating 2072 bytes". 
I have not (yet) been able to pinpoint to any specific piece of coding; it appears to be related to programsize as removing / re-inserting a block of comments (some 30 lines of just text comprising my to-do's) is the difference between "execution" and "MemoryError".
From top to bottom have I inserted gc.collect() /  print(gc.mem-free())'s and I can see the free-space go down to around 158k bytes as variables are declared, class-instances instantiated and somehow that number appears to be a critical threshold for one of the steps in the interpretation of the code. 

nb-1. I am aware that my program can be segmented, set_up as frozen code etc but understanding that there is a practical limit allows to stop figuring out if and where I my coding created this situation; 
nb_2: as always, gradly accept suggestions as to what other causes there could be;
nb_3: running a 2-line program with import gc, print gc.mem_free() after reset returns a value of 221104;
nb_4: the amount given in the MemoryError message is not fixed; it changes when i comment_out line(s) of code;
nb_5: using Rp2040 with MP v1.20.0-498-g5dbd6fc70)

**Comment by smurfix** [comment, marked as answer]
Interpreting Python requires memory, and the resulting bytecode requires memory too. There's no way around it.

Step 1 is to pre-compile the code, i.e. upload as .mpy files instead of .py. They'll still be in RAM because littlefs doesn't support mapping to memory (AFAIK it cannot guarantee that the data is contiguous and most MCUs don't have a memory manager), but at least there's no more overhead for the interpreter.

If that doesn't help enough, you need to freeze them into the MicroPython image, which requires that you build your own. The manual describes how to do that.

**Reply by ctimmer** [reply]
You might take a look at pyminify. It compresses the python source to minimum, like removing all comments, renaming variable names, etc. I haven't had any problems using it with micropython. Could make testing a real nightmare but it has a lot of options.

Curt

**Reply by Flipje1955** [reply]
Matthias, Curt, thank you for your suggestions ! I was expecting to run into the (RAM) limitations of the board at a given moment but was not sure when and how limitations would present themself. Although you do not share any numbers, it appears that code + interpreter etc needs about 160k.
As I am still developing / testing the code, so using a compression tool at this point may in the end cost me more time than using the cross compiler and/or start to freeze consolidated parts. 
Thx again !
